File: Arrays/0014-longest-common-prefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        # Edge case: empty list
        if not strs:
            return ""

        # Take first string as reference prefix
        prefix = strs[0]

        # Compare prefix with each string and shrink it
        for s in strs[1:]:
            # Keep reducing prefix until it matches start of s
            while not s.startswith(prefix):
                prefix = prefix[:-1]
                if not prefix:
                    return ""

        return prefix

File: string/0014-longest-common-prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty()) return "";

        // First string ko reference prefix maan lete hain
        string prefix = strs[0];

        // Rest of the strings ke saath compare karte hain
        for (int i = 1; i < strs.size(); i++) {
            // Jab tak current string `strs[i]` prefix se start na ho
            while (strs[i].find(prefix) != 0) {
                // Prefix ka aakhri character trim karo
                prefix.pop_back(); // Basic Operation: Trimming/Access
                
                if (prefix.empty()) return "";
            }
        }

        return prefix;
    }
};
